api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/CommonQueries.java

```diff
@@ -148,4 +150,39 @@ public static String getPatientsOnTbTreatmentQuery(
 
     return replaced;
   }
+
+  /**
+   * Gets last obs with value coded before enDate
+   *
+   * @param encounterTypeId The Obs encounter Type
+   * @param question The Obs quetion concept
+   * @param answers The third value coded
+   * @return String
+   */
+  public static String getLastCodedObsBeforeDate(
+      Integer encounterTypeId, Integer question, List<Integer> answers) {
+
+    Map<String, String> map = new HashMap<>();
+
+    map.put("encounterTypeId", String.valueOf(encounterTypeId));
+    map.put("question", String.valueOf(question));
+    map.put("answer", StringUtils.join(answers, ","));
+
+    String query =
+        "SELECT p.patient_id FROM patient p INNER JOIN encounter e "
+            + "ON p.patient_id = e.patient_id "
+            + "INNER JOIN obs o "
+            + "ON e.encounter_id = o.encounter_id "
+            + "WHERE e.location_id = :location AND e.encounter_type = ${encounterTypeId} "
+            + "AND o.concept_id = ${question} AND o.value_coded IN (${answers}) "
+            + "AND e.encounter_datetime <= endDate  "
+            + "AND p.voided = 0 AND e.voided = 0 AND o.voided = 0 "
+            + "ORDER BY e.encounter_datetime DESC ";
```

Comment:
Order BY  DESC clause will sort your results starting with the most recent one(latest date), so if the patient has 10 results/encounters, they will be listed in that order to the earliest one, this mean our result set will be very huge with duplicates and what we don't need. What if you use the MAX function on the encounter/Obs with only patient_id and encounter_datime grouped by patient_id the join that results with obs, then later return unique patient_ids, I think that will have impact on perfrmance

Reply:
Looks like this function is NOT used anywhere, of what use is there in the code?

Reply:
@ningosi  Apparently Itelio copied to  GenericCohortQueries by equest from yassin, so will remove on the Common, About the query problably will have some updates that are still on discussion.